MONet Bundle Integration into MONAI Deploy

SimoneBendazzoli93 opened the pull request:
This PR introduces support for the MONet Bundle (an nnUNet wrapper for the MONAI Bundle) into MONAI Deploy.

Key Features:

- Added a new operator: **`MONetBundleInferenceOperator`**, extending **`MonaiBundleInferenceOperator`**

- Included an example application demonstrating spleen segmentation using the **`MONetBundleInferenceOperator`**

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added MONetBundleInferenceOperator — a specialized inference operator with nnUNet-style model support, multimodal input handling, automatic predictor setup, and streamlined prediction behavior.
* **Bug Fixes**
  * Fixed YAML extension handling when reading bundle configurations.
  * Normalized missing input metadata to ensure consistent validation and processing.
* **Chores**
  * Exported MONetBundleInferenceOperator in the package public API.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SimoneBendazzoli93 committed:
Add MONetBundleInferenceOperator to the operator module

- Included MONetBundleInferenceOperator in the __init__.py file for operator registration.
- Updated import statements to reflect the addition of the new operator.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

SimoneBendazzoli93 committed:
Fix minor issues in monai_bundle_inference_operator.py

- Corrected the bundle suffixes tuple to include a period before 'yml'.
- Fixed a method call to ensure casefold() is invoked correctly.
- Initialized meta_data to an empty dictionary if not provided.

These changes enhance code clarity and prevent potential runtime errors.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

SimoneBendazzoli93 committed:
Add MONetBundleInferenceOperator implementation

- Introduced a new operator, MONetBundleInferenceOperator, for performing inference using the MONet bundle.
- Extended functionality from MonaiBundleInferenceOperator to support nnUNet-specific configurations.
- Implemented methods for initializing configurations and performing predictions with multimodal data handling.

This addition enhances the inference capabilities within the MONAI framework.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

SimoneBendazzoli93 committed:
Add monet_bundle_inference_operator.py

- Introduced a new file containing the implementation of the MONetBundleInferenceOperator.
- This operator extends the MonaiBundleInferenceOperator to facilitate inference with nnUNet-specific configurations.
- Implemented methods for configuration initialization and multimodal data prediction, enhancing the MONAI framework's inference capabilities.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

SimoneBendazzoli93 committed:
Add MONetBundleInferenceOperator to __all__ in operators module

- Registered MONetBundleInferenceOperator in the __init__.py file to ensure it is included in the module's public API.
- This change facilitates easier access to the operator for users of the MONAI framework.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

MMelQin committed:
Fix code and improve unit tests to ensure nvimgcodec based decoder is tested alone  (#573)

* Added saving decoded pixels for in deepth review if needed

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Fixed linting complaints

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Fixed the code and improve the tests with failed tests to be addressed.

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Force YBR for JEPG baseline, and test nvimgcodec without any decault decoders

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Critical changes make uncompressed images matching pydicom default decoders.

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Removed support for 12bit "JPEG Extended, Process 2+4"

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Address review comments including from AI agent

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Added reason for ignoring dcm files known to fail to uncompress

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Updated the notes on perf test results

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Explicitly minimized lazy loading impact and added comments on it.

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Updated doc sentences

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Editorial changes made to comments

Signed-off-by: M Q <mingmelvinq@nvidia.com>

---------

Signed-off-by: M Q <mingmelvinq@nvidia.com>
Signed-off-by: Simone Bendazzoli <simben@kth.se>

MMelQin committed:
Mq/release 3.5 (#575)

* Release v3.5.0

Signed-off-by: M Q <mingmelvinq@nvidia.com>

* Bump version: 3.4.0 → 3.5.0

Signed-off-by: M Q <mingmelvinq@nvidia.com>

---------

Signed-off-by: M Q <mingmelvinq@nvidia.com>
Signed-off-by: Simone Bendazzoli <simben@kth.se>

chezhia commented on devel/monet_bundle_inference_operator.py:
This breaks the existing project structure by introducing a new folder 'devel', can this not be part of the file under ''operators" folder?  - Also, can you add some links in the docstrings on how to generate the MONET bundle. Does this currently support all versions of nnunet?

SimoneBendazzoli93 replied:
The devel folder was accidentally included in the PR. I have now removed, adding also some references to the MONet Bundle in the docstrings

SimoneBendazzoli93 committed:
Remove monet_bundle_inference_operator.py file, which contained the implementation of the MONetBundleInferenceOperator. This deletion simplifies the codebase by eliminating unused or redundant components.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

SimoneBendazzoli93 committed:
Add documentation for MONetBundleInferenceOperator

- Enhanced the docstring for MONetBundleInferenceOperator to include a reference to the MONet bundle repository and provide additional context on its functionality.
- This update improves clarity for users regarding the operator's purpose and usage.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

SimoneBendazzoli93 committed:
Refactor MONetBundleInferenceOperator type checking and formatting

- Improved the type checking for the model_network parameter to enhance readability and maintainability.
- Adjusted formatting in the predict method for better clarity and consistency in multimodal data handling.
- These changes contribute to cleaner code and improved functionality within the MONAI framework.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

SimoneBendazzoli93 committed:
Add TritonModel support in MONetBundleInferenceOperator

- Integrated TritonModel type checking into the MONetBundleInferenceOperator to enhance model compatibility.
- Updated the predict method to retain metadata from input data, improving the output structure for predictions.

These changes improve the operator's functionality and usability within the MONAI framework.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

chezhia committed:
Update monai/deploy/operators/monet_bundle_inference_operator.py

Minor typos

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Signed-off-by: Elanchezhian <chezhipower@gmail.com>
Signed-off-by: Simone Bendazzoli <simben@kth.se>

chezhia commented on monai/deploy/operators/monet_bundle_inference_operator.py:
```diff
@@ -0,0 +1,100 @@
+
+        self._nnunet_predictor: torch.nn.Module = None
+
+    def _init_config(self, config_names):
+
+        super()._init_config(config_names)
+        parser = get_bundle_config(str(self._bundle_path), config_names)
+        self._parser = parser
+
+        self._nnunet_predictor = parser.get_parsed_content("network_def")
```
This suggestion from copilot needs review - If the parent class is initializing the model properly, we can accept the suggestion, but I am not sure if get_bundle_config is needed to apply any config patch.

chezhia replied:
@SimoneBendazzoli93  - please review this suggestion above

chezhia committed:
Apply suggestions from code review

Applying minor patch to docs

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Signed-off-by: Elanchezhian <chezhipower@gmail.com>
Signed-off-by: Simone Bendazzoli <simben@kth.se>

chezhia committed:
Update monai/deploy/operators/monet_bundle_inference_operator.py

minor change for stability

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Signed-off-by: Elanchezhian <chezhipower@gmail.com>
Signed-off-by: Simone Bendazzoli <simben@kth.se>

chezhia commented:
@SimoneBendazzoli93  It looks like the DCO (Developer Certificate of Origin) check is failing. To fix this, please ensure all your commits are signed off.

You can do this by amending your previous commits using:
git commit --amend --signoff

Or, if you have multiple commits, you can perform an interactive rebase:
git rebase -i main --signoff

Then, force-push the changes to the branch. This is required for the PR to be mergedr

chezhia commented on monai/deploy/operators/monet_bundle_inference_operator.py:
```diff
@@ -0,0 +1,99 @@
+
+        self._set_model_network(self._model_network)
+
+        if len(kwargs) > 0:
+            multimodal_data = {"image": data}
+            for key in kwargs.keys():
+                if isinstance(kwargs[key], MetaTensor):
+                    multimodal_data[key] = ResampleToMatch(mode="bilinear")(kwargs[key], img_dst=data)
+            data = ConcatItemsd(keys=list(multimodal_data.keys()), name="image")(multimodal_data)["image"]
```
I think this can be ignored

chezhia committed:
Apply suggestion from @coderabbitai[bot]

protection for meta attribute - added safety

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Signed-off-by: Elanchezhian <chezhipower@gmail.com>
Signed-off-by: Simone Bendazzoli <simben@kth.se>

chezhia reviewed (changes requested):
A couple of items identified by copilot needs review from author. Accepted a few minor suggestions.

SimoneBendazzoli93 committed:
Refactor MONetBundleInferenceOperator configuration initialization

- Updated the _init_config method to directly use the parser instance for retrieving the network definition, improving code clarity and reducing redundancy.
- This change enhances the maintainability of the MONetBundleInferenceOperator within the MONAI framework.

Signed-off-by: Simone Bendazzoli <simben@kth.se>

SimoneBendazzoli93 committed:
Merge remote-tracking branch 'upstream/main'

SimoneBendazzoli93 committed:
Merge remote-tracking branch 'upstream/main'
